pkg/controllers/apply_controller.go

```diff
@@ -0,0 +1,356 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/dynamic"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	workv1alpha1 "sigs.k8s.io/work-api/pkg/apis/v1alpha1"
+)
+
+// ApplyWorkReconciler reconciles a Work object
+type ApplyWorkReconciler struct {
+	client             client.Client
+	spokeDynamicClient dynamic.Interface
+	log                logr.Logger
+	restMapper         meta.RESTMapper
+}
+
+type applyResult struct {
+	identifier workv1alpha1.ResourceIdentifier
+	generation int64
+	updated    bool
+	err        error
+}
+
+// Reconcile implement the control loop logic for Work object.
+func (r *ApplyWorkReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	work := &workv1alpha1.Work{}
+	err := r.client.Get(ctx, types.NamespacedName{Name: req.Name, Namespace: req.Namespace}, work)
+	switch {
+	case errors.IsNotFound(err):
+		return ctrl.Result{}, nil
+	case err != nil:
+		return ctrl.Result{}, err
+	}
+
+	// do nothing if the finalizer is not present
+	// it ensures all maintained resources will be cleaned once work is deleted
+	found := false
+	for i := range work.Finalizers {
+		if work.Finalizers[i] == workFinalizer {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return ctrl.Result{}, nil
+	}
```

Comment:
```suggestion
	if !controllerutil.ContainsFinalizer(work, workFinalizer) {
		return ctrl.Result{}, nil
	}
```
Maybe we can use [ContainsFinalizer](https://github.com/kubernetes-sigs/controller-runtime/blob/197751df6040ec99414574e89f3fa73914ce335d/pkg/controller/controllerutil/controllerutil.go#L363) to simply this logic.

Reply:
yeah, this is neat.